File: src/main/java/com/pygostylia/osprey/BoatEntity.java
package com.pygostylia.osprey;

import java.io.IOException;
import java.util.ArrayList;

public class BoatEntity extends ObjectEntity {
    static final int TYPE = Registry.entity("minecraft:boat");

    public boolean turningLeft;
    public boolean turningRight;
    ArrayList<Entity> passengers = new ArrayList<>();

    public BoatEntity(Position position) {
        super(position);
    }

    @Override
    int type() {
        return TYPE;
    }

    @Override
    float colliderXZ() {
        return 1.375f;
    }

    @Override
    float colliderY() {
        return 0.5625f;
    }

    @Override
    public void spawnForPlayer(Player player) {
        super.spawnForPlayer(player);
        if (!passengers.isEmpty()) {
            updatePassengers(player);
        }
    }

    void updatePassengers(Player player) {
        player.sendSetPassengers(this, passengers);
    }

    void updatePassengers() {
        playersWithLoaded.forEach(this::updatePassengers);
    }

    public void addPassenger(Entity passenger) {
        passengers.add(passenger);
        if (passenger instanceof Player player) {
            player.setRidingEntity(this);
        }
        updatePassengers();
    }

    public void removePassenger(Entity passenger) {
        passengers.remove(passenger);
        if (passenger instanceof Player player) {
            player.setNotRidingEntity();
        }
        updatePassengers();
    }

    @Override
    public void interact(Player sender) {
        addPassenger(sender);
    }

    public void dismount(Player sender) {
        removePassenger(sender);
    }

    @Override
    public void attack(Player sender) {
        destroy();
    }
}

File: src/main/scala/com/pygostylia/osprey/BoatEntity.scala
package com.pygostylia.osprey

import java.util

object BoatEntity {
  private[osprey] val TYPE = Registry.entity("minecraft:boat")
}

final class BoatEntity(p: Position) extends ObjectEntity(p) {
  private[osprey] val passengers = new util.ArrayList[Entity]
  var turningLeft = false
  var turningRight = false
  position = p

  override def spawnForPlayer(player: Player): Unit = {
    super.spawnForPlayer(player)
    if (passengers.isEmpty) updatePassengers(player)
  }

  override def interact(sender: Player): Unit = addPassenger(sender)

  def addPassenger(passenger: Entity): Unit = {
    passengers.add(passenger)
    passenger match {
      case player: Player => player.setRidingEntity(this)
    }
    updatePassengers()
  }

  private[osprey] def updatePassengers(): Unit =
    playersWithLoaded.forEach(this.updatePassengers)

  private[osprey] def updatePassengers(player: Player): Unit =
    player.sendSetPassengers(this, passengers)

  def dismount(sender: Player): Unit = removePassenger(sender)

  def removePassenger(passenger: Entity): Unit = {
    passengers.remove(passenger)
    passenger match {
      case player: Player => player.setNotRidingEntity()
    }
    updatePassengers()
  }

  override def attack(sender: Player): Unit = destroy()

  override private[osprey] def `type` = BoatEntity.TYPE

  override private[osprey] def colliderXZ = 1.375f

  override private[osprey] def colliderY = 0.5625f
}
